nano/node/confirmation_height_processor.cpp

```diff
@@ -0,0 +1,372 @@
+#include <nano/node/confirmation_height_processor.hpp>
+
+#include <boost/optional.hpp>
+#include <cassert>
+#include <nano/lib/logger_mt.hpp>
+#include <nano/lib/numbers.hpp>
+#include <nano/lib/utility.hpp>
+#include <nano/node/active_transactions.hpp>
+#include <nano/node/stats.hpp>
+#include <nano/secure/blockstore.hpp>
+#include <nano/secure/common.hpp>
+#include <nano/secure/ledger.hpp>
+#include <numeric>
+
+nano::confirmation_height_processor::confirmation_height_processor (nano::pending_confirmation_height & pending_confirmation_height, nano::block_store & store, nano::ledger & ledger, nano::active_transactions & active, nano::logger_mt & logger) :
+pending_confirmations (pending_confirmation_height),
+store (store),
+ledger (ledger),
+active (active),
+logger (logger),
+thread ([this]() {
+	nano::thread_role::set (nano::thread_role::name::confirmation_height_processing);
+	this->run ();
+})
+{
+}
+
+nano::confirmation_height_processor::~confirmation_height_processor ()
+{
+	stop ();
+}
+
+void nano::confirmation_height_processor::stop ()
+{
+	stopped = true;
+	condition.notify_one ();
+	if (thread.joinable ())
+	{
+		thread.join ();
+	}
+}
+
+void nano::confirmation_height_processor::run ()
+{
+	std::unique_lock<std::mutex> lk (pending_confirmations.mutex);
+	while (!stopped)
+	{
+		if (!pending_confirmations.pending.empty ())
+		{
+			current_original_pending_block = *pending_confirmations.pending.begin ();
+			pending_confirmations.pending.erase (current_original_pending_block);
+			lk.unlock ();
+			add_confirmation_height (current_original_pending_block);
+			current_original_pending_block = 0;
```

Comment:
Is it really safe to update this member outside the lock?

Reply:
oh gosh no!! That needs to be protected, good spot. I haven't run TSAN on the most recent set of changes.